### PR TITLE
[FIX] mail: fix missing space between the task name and calendar icon 

### PR DESCRIPTION
**Pull request by shup-odoo**
Steps to reproduce:
-----------------------------------------------------------------------------------------------------------
1. Install the Project.
2. Go to settings and select customer ratings.
3. Then go to any project edit the stage and select the template for the rating
 email template.
4. Then change the stage of any task. check the email and give a rating to your
 task.
5. Check the Discus app then you will see a notification and no space between the
 name of the task and the calendar icon. 

Issue:
-----------------------------------------------------------------------------------------------------------
- A missing space between the name of the task and the calendar icon. 

Cause:
-----------------------------------------------------------------------------------------------------------
- The task's name and the calendar icon are not separated.

Fix:
-----------------------------------------------------------------------------------------------------------
- After this commit there is a space between the name of the task and the
 calendar icon 

task-3681318

### CODE REVIEW

**Review by pkyriakou — COMMENTED** [review]
Thank you @shup-odoo LGTM :100:

**Comment by xavierbol** [comment]
@shup-odoo can you check if the issue is still there in 17? If yes, can you retarget your fix? 🙂 

Thanks in advance!

**Comment by shup-odoo** [comment]
Hello @xavierbol 
Made the changes, please have a look.
Thanks !!

File: addons/mail/static/src/core/common/message.xml [hunk 1]
```diff
@@ -55,7 +55,7 @@
                             </small>
                             <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
                                 <span class="text-600 cursor-pointer">
-                                    <i class="fa fa-calendar-o"/>
+                                    <i class="fa fa-calendar-o ms-1"/>
```
Comment (ppr-odoo) [inline addons/mail/static/src/core/common/message.xml:58]: According to the this [line](https://github.com/odoo/odoo/pull/151700/files#diff-0fc14194d473213159537bf41f421dcbaf940dfd81ac87d263b196e914d40af3R56), extra space will be added to the chat window.

File: addons/mail/static/src/core/common/message.xml [hunk 2]
```diff
@@ -45,7 +45,7 @@
                                 className="'ms-1'"
                                 message="props.message"
                                 thread="props.thread"/>
-                            <small t-if="!isOriginThread and !message.is_transient" t-on-click.prevent="openRecord" class="ms-1 text-500">
+                            <small t-if="!isOriginThread and !message.is_transient" t-on-click.prevent="openRecord" class="mx-1 text-500">
```
Comment (ppr-odoo) [inline addons/mail/static/src/core/common/message.xml:48]:
```suggestion
                            <small t-if="!isOriginThread and !message.is_transient" t-on-click.prevent="openRecord" class="ms-1 me-1 text-500">
```
We are violating our stability policy.

**Comment by ppr-odoo** [comment]
> Actually, according to line 56
> 
> ```
> <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
> ```
> 
> we already add a space before the calendar icon. 🤔



> Actually, according to line 56
> 
> ```
> <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
> ```
> 
> we already add a space before the calendar icon. 🤔

Currently, the margin is only set for the chat window. However, in our case, it is the Discuss app's inbox. The chat window and inbox both are different.

`<div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight or props.message.scheduledDate }" t-att-title="messageService.scheduledDateSimple(props.message)">`

Is this approach acceptable?

**Comment by xavierbol** [comment]
> > Actually, according to line 56
> > ```
> > <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we already add a space before the calendar icon. 🤔
> 
> > Actually, according to line 56
> > ```
> > <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we already add a space before the calendar icon. 🤔
> 
> Currently, the margin is only set for the chat window. However, in our case, it is the Discuss app's inbox. The chat window and inbox both are different.
> 
> `<div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight or props.message.scheduledDate }" t-att-title="messageService.scheduledDateSimple(props.message)">`
> 
> Is this approach acceptable?

Not really since `props.message.scheduledDate` will be always true. Your suggestion seems to be equivalent to:

`<div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow }" t-att-title="messageService.scheduledDateSimple(props.message)">`

TBH, I don't know if it is a good idea, we will ask a review to Discuss team to know what they think once you will have done the changes. 🙂

**Comment by shup-odoo** [comment]
Hello @xavierbol 

I believe that `props.isInChatWindow` is True only when in the chat window, not in the Discuss app's inbox and `props.message.scheduledDate` will not always be True. It will be True in scenarios where we are rating and seeing the message notification in the Discuss app's inbox. In that case, we can add a condition as @ppr-odoo  suggested for the margin like `props.message.scheduledDate or (props.isInChatWindow and isAlignedRight)`. please let me know if i have missed something here.

Thank you !!

**Comment by xavierbol** [comment]
> Hello @xavierbol
> 
> I believe that `props.isInChatWindow` is True only when in the chat window, not in the Discuss app's inbox and `props.message.scheduledDate` will not always be True. It will be True in scenarios where we are rating and seeing the message notification in the Discuss app's inbox. In that case, we can add a condition as @ppr-odoo suggested for the margin like `props.message.scheduledDate or (props.isInChatWindow and isAlignedRight)`. please let me know if i have missed something here.
> 
> Thank you !!

The problem in that suggestion is `props.message.scheduledDate` will be always `true` since it is used in `t-if` to know if we have to display or not the div.

So, we have to check if the margin is needed in all cases or not. 🙂

**Comment by shup-odoo** [comment]
Hello @xavierbol 

I apologize for the inconvenience :disappointed: . I initially couldn’t determine the cause of the issue, but after further investigation, I found a solution that has currently implemented.

Here’s how it works: When we send a rating request email to a customer, they click an emoji face to rate their experience. However, if the customer doesn't submit the feedback form, the initial rating is still generated in the chat (in this case, the calendar icon will show). If the customer leaves the page without submitting the rating, the initial rating will be recorded. The customer has 2 hours to rate and submit the feedback form; otherwise, the initial rating will be submitted.

I have attached PR and some screenshots for your reference.

When a user is a customer and they give a rating, the conditions props.isInChatWindow and isAlignedRight (in the chat window) will be true.
Screenshot from 2024-07-25 14-21-23

When a customer gives a rating, the condition !isAlignedRight (in the chat window) will be true.
Screenshot from 2024-07-25 14-22-03

In our case, we are adding the margin-end , which will add space before the calendar icon.
Screenshot from 2024-07-25 14-28-27

PR LINK - https://github.com/odoo/odoo/pull/95623

Have a nice day and Thank you :smiley:

**Comment by xavierbol** [comment]
Hello @shup-odoo @ppr-odoo ,

The problem with your current solution is we will always add the margin right and so as you can see we have more space between the display name of the record and the calendar icon in the 2 first screenshots you shared in your last comment.

I really think we should change the condition to add ms-2 in our use case as well.

File: addons/mail/static/src/core/common/message.xml [hunk 3]
```diff
@@ -36,7 +36,7 @@
                                 <strong class="me-1"><t t-if="message.author" t-esc="message.author.name"/><t t-else="" t-esc="message.email_from"/></strong>
                             </span>
                             <t t-if="!isAlignedRight" t-call="mail.Message.notification"/>
-                            <small t-if="!message.is_transient" class="o-mail-Message-date text-muted opacity-75" t-att-class="{ 'me-2': !isAlignedRight }" t-att-title="message.datetimeShort">
+                            <small t-if="!message.is_transient" class="o-mail-Message-date text-muted opacity-75" t-att-title="message.datetimeShort">
```
Comment (xavierbol) [inline addons/mail/static/src/core/common/message.xml:39]: I still don't know why you remove that, to me, it does not seem related to the issue to fix. 🤔 

May I miss something?

Comment (shup-odoo) [reply addons/mail/static/src/core/common/message.xml:39]: Hello @xavierbol 

Since we are adding a space in every scenario I think we should remove it to avoid the extra space. If i miss something please let me know 😃

Comment (xavierbol) [reply addons/mail/static/src/core/common/message.xml:39]: No, I don't agree, it is not related to the issue we try to fix.

File: addons/mail/static/src/core/common/message.xml [hunk 4]
```diff
@@ -53,7 +53,7 @@
                                     (from <a t-att-href="message.resUrl"><t t-esc="message.originThread?.prefix"/><t t-esc="message.originThread?.displayName"/></a>)
                                 </t>
                             </small>
-                            <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
+                            <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight or (!isOriginThread and !message.is_transient) }" t-att-title="messageService.scheduledDateSimple(props.message)">
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message.xml:56]: We will likely forget to adapt this condition is we ever adapt the other one. Maybe there should be a common getter for `hasThreadName` (or something like that).

Comment (seb-odoo) [inline addons/mail/static/src/core/common/message.xml:56]: Also I would add parenthesis on the left part of the `or` too as I'm always confused with priority (also to be consistent as you add them on the right part).

**Review by seb-odoo — COMMENTED** [review]
Hello, a before/after screenshot in the PR would help, but it looks like an acceptable change.

File: addons/mail/static/src/core/common/message.xml (hunk 4 above)
Comment (shup-odoo) [reply addons/mail/static/src/core/common/message.xml]: Hello @seb-odoo

Thanks for your feedback 😃 since `hasThreadName` was removed in saas-16.3 I checked for alternatives but didn't find any alternative so I created a new method.

File: addons/mail/static/src/core/common/message.js [hunk 5]
```diff
@@ -307,6 +307,10 @@ export class Message extends Component {
         return Boolean(this.env.inChatWindow && this.props.message.isSelfAuthored);
     }
 
+    get isNotOriginThreadAndIsNotTransient() {
+        return !this.isOriginThread && this.props.hasActions;
+    }
+
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/message.js:313]: `AndIsNotTransient` comes from previous code that was `and !message.is_transient` but this has been replaced with `this.props.hasActions`. Is it intentional?

`hasActions` is an option to show actions (when available) on the message. It doesn't mean that the message has no actions (e.g. as a consequence of being transient). For example, the Dialog shown to confirm message deletion shows the same message without the actions. That's one use-case of `props.hasActions = false`.

I would name the getter `isPersistentMessageFromAnotherThread`: it's easier to read when there's no `"Not"` (twice)

File: addons/mail/static/src/core/common/message.xml [hunk 6]
```diff
@@ -45,15 +45,15 @@
...
+                            <small t-if="isNotOriginThreadAndIsNotTransient" t-on-click.prevent="openRecord" class="ms-1 text-500">
                                 <t t-if="message.model !== 'discuss.channel'">
                                     on <a t-att-href="message.resUrl"><t t-esc="message.originThread?.displayName"/></a>
                                 </t>
                                 <t t-if="message.model === 'discuss.channel'">
                                     (from <a t-att-href="message.resUrl"><t t-esc="message.originThread?.prefix"/><t t-esc="message.originThread?.displayName"/></a>)
                                 </t>
                             </small>
-                            <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': props.isInChatWindow and isAlignedRight }" t-att-title="messageService.scheduledDateSimple(props.message)">
+                            <div t-if="props.message.scheduledDate" t-att-class="{ 'ms-2': (props.isInChatWindow and isAlignedRight) or (isNotOriginThreadAndIsNotTransient)}" t-att-title="messageService.scheduledDateSimple(props.message)">
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/message.xml:56]: Nitpick: keep whitespace before `}`, since we have one after `{`. The whitespace are preferable for readability

**Comment by shup-odoo** [comment]
Hello @xavierbol
I have made changes, please have a look.
Thank you 😄

**Comment by xavierbol** [comment]
Thanks @suth-odoo 

I guess now we can merge that fix. 🙂 

Thanks @seb-odoo @alexkuhn for your reviews! 

robodoo r+
